### PR TITLE
Feature/society join

### PR DESCRIPTION
**Pull request by Plebbaroni**
This is what I did:

- Join/Leave endpoints for societies(links a user to a society and vice versa)
- Attend/Unattend endpoints for events(works the same way as the society endpoints)
- Used try-catch blocks for some previous endpoints(create society/event)

### CODE REVIEW

File: backend/prisma/schema.prisma [hunk 1]
```diff
@@ -19,55 +19,51 @@ enum UserType {
...
 model Keyword {
-  id Int @id @default(autoincrement())
-  text String @unique
-  attendees Attendee[]
-  events Event[] 
-}
\ No newline at end of file
+  id        Int     @id @default(autoincrement())
+  text      String  @unique
+  attendees User[]
```
Comment (lachlanshoesmith) [inline backend/prisma/schema.prisma:67]: i would suggest changing the name of this to 'subscribers' maybe, might be a bit clearer. confused me for a bit

File: backend/prisma/schema.prisma [hunk 2]
```diff
@@ -19,55 +19,51 @@ enum UserType {
...
-  society Society @relation(fields:[societyId], references: [id], onDelete: Cascade)
-  keywords Keyword[]
+  endDateTime   DateTime
+  location      String
+  description   String
+  societyId     Int
+  society       Society   @relation(fields: [societyId], references: [id], onDelete: Cascade)
+  keywords      Keyword[]
+  attendees     User[]
+  numAttendees  Int       @default(0)
```
Comment (lachlanshoesmith) [inline backend/prisma/schema.prisma:59]: unless this is something prisma specifically suggests i would avoid this personally. when we add or remove attendees we'd have to update this value as well and that either sounds:

1. bound to cause concurrency problems, or
2. issues where one is updated and not the other by mistake

File: backend/prisma/schema.prisma [hunk 3]
```diff
@@ -19,55 +19,51 @@ enum UserType {
...
+  members        User[]  @relation(name: "membertosociety")
+  discordId      String? @unique
   profilePicture String?
 }
 
 model Event {
-  numId Int @id @default(autoincrement())
-  banner String
-  name String
+  numId         Int       @id @default(autoincrement())
```
Comment (lachlanshoesmith) [inline backend/prisma/schema.prisma:48]: could this just be called eventId or similar? Or even id?

File: backend/src/index.ts [hunk 4]
```diff
@@ -66,6 +74,8 @@ app.get("/", (req: Request, res: Response) => {
   res.send("Hello, TypeScript with Express :)))!");
 });
 
+//Will refactor this soon, I think prisma throws its own error when some
```
Comment (lachlanshoesmith) [inline backend/src/index.ts:77]: backend validation separate from the database is a good idea for security reasons amongst other things. i'd recommend both, or at least certainly keeping it on the backend

File: backend/src/index.ts [hunk 5]
```diff
@@ -296,77 +306,98 @@ app.get("/user", async (req, res: Response) => {
...
-  const newSociety = await prisma.society.create({
-    data: {
-      name: society.name,
-      admin: {
-        connect: {
-          id: society.userId,
-        }
-      }
+    if (!society.profilePicture) {
+      society.profilePicture = "null";
```
Comment (lachlanshoesmith) [inline backend/src/index.ts:318]: does prisma let you just use `null`?

File: backend/src/index.ts [hunk 6]
```diff
@@ -296,77 +306,98 @@ app.get("/user", async (req, res: Response) => {
...
+    }
+  }
+);
+
+function isValidDate(startDate: Date, endDate: Date): boolean {
   //Should probably add more test cases here
-  var parsedStartDate = dayjs(startDate)
-  var parsedEndDate = dayjs(endDate)
-  //Might need to keep an eye on this condition, I'm subtracting two mins from dayjs as allowance as 
+  var parsedStartDate = dayjs(startDate);
```
Comment (lachlanshoesmith) [inline backend/src/index.ts:390]: please use `const`/`let` instead of `var` throughout the program, including tests

File: backend/src/index.ts [hunk 7]
```diff
@@ -296,77 +306,98 @@ app.get("/user", async (req, res: Response) => {
...
+  var parsedEndDate = dayjs(endDate);
+  //Might need to keep an eye on this condition, I'm subtracting two mins from dayjs as allowance as
   //previously it registered that dayJs() is after startDate.
-  return (!(parsedStartDate.isAfter(parsedEndDate) || parsedStartDate.isSame(parsedEndDate) || parsedStartDate.isBefore(dayjs().subtract(2, 'minutes'))))
+  return !(
+    parsedStartDate.isAfter(parsedEndDate) ||
+    parsedStartDate.isSame(parsedEndDate) ||
+    parsedStartDate.isBefore(dayjs().subtract(2, "minutes"))
```
Comment (lachlanshoesmith) [inline backend/src/index.ts:397]: the 2 min before bit seems kind of arbitrary, i'd remove it for clarity

File: backend/src/index.ts [hunk 8]
```diff
@@ -402,6 +433,186 @@ app.get("/societies", async (req, res: Response) => {
...
+    where: {
+      numId: event.numId,
+    },
+    data: {
+      attendees: {
+        disconnect: {
+          id: userID,
+        },
+      },
+      numAttendees: {
```
Comment (lachlanshoesmith) [inline backend/src/index.ts:600]: yeah.. this seems like it's a bit of a hassle to manage

File: backend/src/index.ts [hunk 9]
```diff
@@ -402,6 +433,186 @@ app.get("/societies", async (req, res: Response) => {
...
+      members: {
+        disconnect: {
+          id: userID,
+        },
+      },
+    },
+  });
+
+  //Not exactly sure what we'd want to be returning here
+  return res.status(200).json({message: "ok"});
```
Comment (lachlanshoesmith) [inline backend/src/index.ts:518]: this is good

File: backend/src/index.ts [hunk 10]
```diff
@@ -402,6 +433,186 @@ app.get("/societies", async (req, res: Response) => {
...
+          },
+        },
+      },
+    });
+
+    return res.status(200).json({message: "ok"});
+  }
+);
+
+app.post("/user/society/leave", async (req: TypedRequest<societyIdBody>, res: Response) => {
```
Comment (lachlanshoesmith) [inline backend/src/index.ts:481]: consider `app.delete('/user/society')`

**Review by lachlanshoesmith — CHANGES_REQUESTED** [review]
very nice work overall!

File: backend/src/index.ts (hunk 5 above)
Comment (Plebbaroni) [reply backend/src/index.ts:318]: I believe so, it's showing up as NULL on supabase. I could change it to smth safer though if needed(maybe setting a default value in the schema could be a better idea?).

File: backend/src/index.ts (hunk 10 above)
Comment (Plebbaroni) [reply backend/src/index.ts:481]: I'll also do the same for user/event/unattend.

File: backend/src/index.ts (hunk 5 above)
Comment (lachlanshoesmith) [reply backend/src/index.ts:318]: to clarify, do you mean `null` shows up as `NULL` or does `"null"` also translate to `NULL`?

Comment (Plebbaroni) [reply backend/src/index.ts:318]: It looks like ```"null"``` is translating to ```NULL```. 

This is what it looks like on supabase:
Screenshot 2024-12-04 at 9 00 40 PM

Comment (lachlanshoesmith) [reply backend/src/index.ts:318]: interesting, would you mind trying `null` on its own?

Comment (Plebbaroni) [reply backend/src/index.ts:318]: ```null``` translates to ```NULL``` if entered as an input on insomnia. If I remove the quotation marks in index.ts typescript gets mad.

Comment (lachlanshoesmith) [reply backend/src/index.ts:318]: what happens if you open `backend/tsconfig.json` and set `strictNullChecks` to `false`?

Comment (Plebbaroni) [reply backend/src/index.ts:318]: Oh nice, it's not angry anymore. I'll keep it as ```null```.

**Review by lachlanshoesmith — APPROVED** [review]

**Pull request merged**